cli/src/bin/phylum.rs

```diff
@@ -180,45 +181,68 @@ async fn handle_commands() -> CommandResult {
         .await;
     }
 
-    let mut api = PhylumApi::new(
-        &mut config.auth_info,
-        &config.connection.uri,
-        timeout,
-        ignore_certs,
-    )
-    .await
-    .context("Error creating client")?;
-
-    // PhylumApi may have had to log in, updating the auth info so we should save the config
-    save_config(&config_path, &config).with_context(|| {
-        format!(
-            "Failed to save configuration to '{}'",
-            config_path.to_string_lossy()
-        )
-    })?;
+    // This is `Future<Output = PhylumApi>`. Commands that require access to the API
+    // will await on this, so that the API is not instantiated ahead of time for subcommands
+    // that don't require it.
+    let api = api_factory(&mut config, &config_path, timeout, ignore_certs);
+
+    let (subcommand, matches) = matches.subcommand().unwrap();
+    match subcommand {
+        "version" => handle_version(&app_name, ver),
+        "update" => handle_update(matches).await,
+        "parse" => handle_parse(matches),
+        "ping" => handle_ping(api.await?).await,
+        "project" => handle_project(&mut api.await?, matches).await,
+        "package" => handle_get_package(&mut api.await?, &config.request_type, matches).await,
+        "history" => handle_submission(&mut api.await?, config, matches).await,
+        "group" => handle_group(&mut api.await?, matches).await,
+        "analyze" | "batch" => handle_submission(&mut api.await?, config, matches).await,
+
+        #[cfg(feature = "selfmanage")]
+        "uninstall" => handle_uninstall(matches),
+
+        #[cfg(feature = "extensions")]
+        "extensions" => handle_extensions(matches).await,
+
+        #[cfg(feature = "extensions")]
+        extension_subcmd => handle_extension_run(extension_subcmd).await,
+
+        #[cfg(not(feature = "extensions"))]
+        _ => unreachable!(),
+    }
+}
 
-    if matches.subcommand_matches("ping").is_some() {
-        let resp = api.ping().await;
-        print_response(&resp, true, None);
-        return Ok(ExitCode::Ok.into());
+#[cfg(feature = "extensions")]
+async fn handle_extension_run(extension_subcmd: &str) -> CommandResult {
+    for extension in extensions::installed_extensions().unwrap_or_default() {
+        if extension_subcmd == extension.name() {
+            return extension.run().await;
+        }
     }
 
-    let should_submit = matches.subcommand_matches("analyze").is_some()
-        || matches.subcommand_matches("batch").is_some();
+    Err(anyhow!("Extension {extension_subcmd} not found"))
```

Comment:
I think this is unreachable. And if we do get here, it's probably because somebody added a new command to `app.rs` and forgot to handle it in `handle_commands()`. In which case, this error message would be confusing. I think a panic is appropriate.

```suggestion
    unreachable!()
```

Reply:
Either way the application will error out. I see no harm in printing an error message that at least attempts to provide some detail.

Reply:
That's fair. But as you [note elsewhere](https://github.com/phylum-dev/cli/pull/415/files/9cf4b7420f079ac9e9017e00b8d58fa11ce2a5a2#r881990505), the error message here is misleading. Using `unreachable!()` is just my way of avoiding writing a better error message 😄